### ISSUE TITLE
Guest fails to reboot if a Network device is attached

### ISSUE THREAD

**Issue by praveen-pk** [cloud-hypervisor/libvirt#65]
While using the following guest definition, guest fails to reboot:

```
<domain type='ch'>
        <name>ch_net_reboot</name>
        <iothreads>3</iothreads>
  <title>Test Cloud Hypervisor Networking</title>
  <description>Network</description>
  <os>
          <type>hvm</type>
     <kernel>hypervisor-fw</kernel>
  </os>
  <memory unit='G'>4</memory>
  <devices>
    <disk type='file'>
            <source file='focal-server-cloudimg-amd64-custom.raw'/>
            <driver name='qemu' if='virtio' type='raw' iothread='1'/>
      <target dev='vda'/>
    </disk>
    <disk type='file'>
            <source file='cloud-config-dhcp.raw'/>
            <driver queues='4'/>
      <target dev='vdb'/>
    </disk>
    <console type='pty'>
      <target type='virtio' port='0'/>
   </console>
   <interface type='network'>
        <mac address='00:11:22:33:44:55'/>
        <driver queues='4'/>
      <source network='default' bridge='virbr0'/>
      <model type='virtio'/>
    </interface>
</devices>
  <vcpu>4</vcpu>
  <cputune>
        <emulatorpin cpuset="0-2"/>
  </cputune>
</domain>
```

Guest reboots just fine when there is no network interface in the definition. From what I see, the vnetx interface created by libvirt gets deleted in the course of guest reboot and cloud-hypervisor cannot boot the guest without the interface.

I confirmed guest reboots just fine while using qemu. While using qemu, the vnetx interface doesn't get deleted.

**Comment by praveen-pk** [comment]
Following are the related journal logs I am seeing

```
Mar 23 06:58:05 praween-ubuntu NetworkManager[813]: <info>  [1616507885.8843] device (virbr0): bridge port vnet0 was attached
Mar 23 06:58:05 praween-ubuntu NetworkManager[813]: <info>  [1616507885.8843] device (vnet0): Activation: connection 'vnet0' enslaved, continuing activation
Mar 23 06:58:05 praween-ubuntu NetworkManager[813]: <info>  [1616507885.8844] device (vnet0): state change: ip-config -> ip-check (reason 'none', sys-iface-state: 'external')
Mar 23 06:58:05 praween-ubuntu NetworkManager[813]: <info>  [1616507885.8947] device (vnet0): state change: ip-check -> secondaries (reason 'none', sys-iface-state: 'external')
Mar 23 06:58:05 praween-ubuntu NetworkManager[813]: <info>  [1616507885.8948] device (vnet0): state change: secondaries -> activated (reason 'none', sys-iface-state: 'external')
Mar 23 06:58:05 praween-ubuntu NetworkManager[813]: <info>  [1616507885.8955] device (vnet0): Activation: successful, device activated.
Mar 23 06:58:07 praween-ubuntu kernel: virbr0: port 2(vnet0) entered learning state
Mar 23 06:58:07 praween-ubuntu avahi-daemon[810]: Joining mDNS multicast group on interface vnet0.IPv6 with address fe80::fc11:22ff:fe33:4455.
Mar 23 06:58:07 praween-ubuntu avahi-daemon[810]: New relevant interface vnet0.IPv6 for mDNS.
Mar 23 06:58:07 praween-ubuntu avahi-daemon[810]: Registering new address record for fe80::fc11:22ff:fe33:4455 on vnet0.*.
Mar 23 06:58:09 praween-ubuntu kernel: virbr0: port 2(vnet0) entered forwarding state
Mar 23 06:59:17 praween-ubuntu avahi-daemon[810]: Interface vnet0.IPv6 no longer relevant for mDNS.
Mar 23 06:59:17 praween-ubuntu kernel: virbr0: port 2(vnet0) entered disabled state
Mar 23 06:59:17 praween-ubuntu kernel: device vnet0 left promiscuous mode
Mar 23 06:59:17 praween-ubuntu kernel: virbr0: port 2(vnet0) entered disabled state
Mar 23 06:59:17 praween-ubuntu audit: ANOM_PROMISCUOUS dev=vnet0 prom=0 old_prom=256 auid=1000 uid=0 gid=0 ses=5
Mar 23 06:59:17 praween-ubuntu avahi-daemon[810]: Leaving mDNS multicast group on interface vnet0.IPv6 with address fe80::fc11:22ff:fe33:4455.
Mar 23 06:59:17 praween-ubuntu NetworkManager[813]: <info>  [1616507957.5697] device (vnet0): state change: activated -> unmanaged (reason 'unmanaged', sys-iface-state: 'removed')
Mar 23 06:59:17 praween-ubuntu avahi-daemon[810]: Withdrawing address record for fe80::fc11:22ff:fe33:4455 on vnet0.
Mar 23 06:59:17 praween-ubuntu NetworkManager[813]: <info>  [1616507957.5701] device (vnet0): released from master device virbr0
```

**Comment by praveen-pk** [comment]
I confirmed it is the kernel that is deleting vnetxx TAP device during guest reboot.

In the current implementation libvirt creates a TAP device and passes the FDs (per queue) to cloud-hypervisor. During guest reboot, cloud-hypervisor closes all FDs that it has open. This triggers clean up on kernel side. 

If all FDs of a TAP device are closed, the underlying device will be closed by kernel. Reference: https://www.kernel.org/doc/Documentation/networking/tuntap.txt

Excerpt:
> A network
>   device will appear as tunXX or tapXX, depending on the options chosen. When
>   the program closes the file descriptor, the network device and all
>   corresponding routes will disappear.


The options to pursue:
* Instead of FDs, pass TAP interface Name to cloud-hypervisor. Cloud-hypervisor should use the TAP interface created by libvirt. Implement the TAP device cleanup on libvirt side, after guest shutdown/Off event.
* Cloud-hypervisor should preserve TAP FDs at the time of guest reboot.

**Comment by sboeuf** [comment]
>In the current implementation libvirt creates a TAP device and passes the FDs (per queue) to cloud-hypervisor. During guest reboot, cloud-hypervisor closes all FDs that it has open. This triggers clean up on kernel side.

Any chance you can modify the libvirt code doing that? If you could `dup()` the file descriptor before passing it to Cloud Hypervisor, this would maintain the TAP opened, and the kernel wouldn't remove it.

>Instead of FDs, pass TAP interface Name to cloud-hypervisor. Cloud-hypervisor should use the TAP interface created by libvirt. Implement the TAP device cleanup on libvirt side, after guest shutdown/Off event.

If you can achieve that and if there's no downside, that sounds good to me.

>Cloud-hypervisor should preserve TAP FDs at the time of guest reboot.

Keeping file descriptors opened across reboots is hard, as we would need a codepath for storing them. And that's not only about keeping an integer, it's about keeping the File or the RawFd.
We've already run into this situation, so maybe we could think about adding a layer between the configuration and the runtime data, in order to store what we need.

**Comment by praveen-pk** [comment]
I will try using dup() before passing FDs to CLH. If that doesn't work, I will attempt passing a TAP device name.

If both options don't work we can pursue storing FDs on CLH side.

**Comment by praveen-pk** [comment]
Using fud() on libvirt before passing the FDs to CLH could get complicated. In this mechanism, libvirt has to store FDs, and re-pass the FDs to CLH during a guest reboot. CLH currently only emits shutdown event, and no reboot events for libvirt to respond.

Instead I chose to pass TAP Name to CLH. This works just fine. PR: https://github.com/cloud-hypervisor/libvirt/pull/66

**Comment by sboeuf** [comment]
>Instead I chose to pass TAP Name to CLH. This works just fine. PR: #66

Sounds good to me!

**Comment by bryteise** [comment]
Fixed via #66. We'll potentially need to revisit this if we want to be able to support additional device types again.
